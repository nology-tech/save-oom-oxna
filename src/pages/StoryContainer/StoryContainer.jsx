import { useNavigate } from "react-router";
import "./StoryContainer.scss";

import { useState } from "react";
import Button from "../../components/Button/Button";
import Layout from "../../components/Layout/Layout";
import storyData from "../../data/storyData";

const StoryContainer = () => {
  const [counter, setCounter] = useState(0);
  const navigate = useNavigate();

  const handleIncrement = () => {
    if (counter > 4) {
      navigate("/swing-game/play");
    } else {
      setCounter(counter + 1);
    }
  };

  return (
    <Layout>
      <div className="story-container">
        <p className="story-container__text">{storyData[counter].story}</p>
        <Button label={"NEXT"} onClick={handleIncrement} />
      </div>
    </Layout>
  );
};

export default StoryContainer;
