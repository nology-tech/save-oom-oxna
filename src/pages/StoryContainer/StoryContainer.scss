@use '../../styles/variables.scss' as *;

.story__container {
  background-image: url(../../assets/images/background.png);
  background-size: cover;
  height: auto;
  width: 100vw;
  background-position: center;
  background-repeat: no-repeat;
  text-align: center;
}
