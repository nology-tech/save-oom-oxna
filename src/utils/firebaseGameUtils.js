import { db } from "../firebase";
import {
  addDoc,
  collection,
  doc,
  getDocs,
  getDoc,
  query,
  where,
  orderBy,
  Timestamp,
} from "firebase/firestore";


/**
 * Gets a user object using the userId.
 * @param {*} userId 
 * @returns 
 */
export const getUserById = async( userId ) => {
  const docRef = doc(db, "users", userId);
  const docSnap = await getDoc( docRef );
  return docSnap;
}

/** 
 * Get all the rounds of a specific game for a given user, ordered by most recent first 
 @param {*} userId
*@returns a querySnap (document snapshot), returns a Promise.
*/

export const getGameRoundsForUser = async (id, game) => {
  const q = query(
    collection(db, "users", id, "rounds_played"),
    where("gameId", "==", game),
    orderBy("roundTime", "desc")
  );
  const querySnap = await getDocs(q);
  return querySnap;
};

/** 
   * Get all the correct round of a specific game for a given user, ordered by most recent first 
  @param {String} userId
  @param {String} gameId
  *@returns a promise with the query snapshot.
  */
export const getCorrectGameRoundsForUser = async (id, game) => {
  const q = query(
    collection(db, "users", id, "rounds_played"),
    where("gameId", "==", game),
    where("correct", "==", true),
    orderBy("roundTime", "desc")
  );
  const querySnap = await getDocs(q);
  return querySnap;
};

/** Get all the incorrect round of a specific game for a given user, ordered by most recent first 
  @param {String} userId
  @param {String} gameId
  *@returns  a promise with the query snapshot.
  */
export const getIncorrectGameRoundsForUser = async (id, game) => {
  const q = query(
    collection(db, "users", id, "rounds_played"),
    where("gameId", "==", game),
    where("correct", "==", false),
    orderBy("roundTime", "desc")
  );
  const querySnap = await getDocs(q);
  return querySnap;
};

/** Function that returns an array of rounds 
  @param {String} userId
  @param {String} gameId
  @param {Function} getGameRounds
  *@returns returns a promise.
  */
export const getArrayOfRounds = async (id, game, getGameRounds) => {
  const querySnap = await getGameRounds(id, game);
  const phonicsArr = querySnap.docs.map((doc) => {
    return doc.data().phonic;
  });
  return phonicsArr;
};

/** Function that returns an array of rounds 
  @param {String} id
  @param {String} game
  @param {Object} roundInfo
  @param {String} answer
  *@returns returs a promise.
  */
export const saveUserRound = async (id, game, level, gameStats, answer) => {
  const { isCorrect, score } = gameStats;
  await addDoc(collection(db, "users", id, "rounds_played"), {
    correct: isCorrect,
    gameId: game,
    phonic: answer,
    level: level,
    roundTime: Timestamp.fromDate(new Date()).toDate(),
    score: score,
  });
};
