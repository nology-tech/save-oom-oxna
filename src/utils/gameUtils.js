import phonicsData from "../data/phonicsData";
import { 
    // getGameRoundsForUser, 
    getCorrectGameRoundsForUser, 
    getIncorrectGameRoundsForUser,
    getArrayOfRounds
     } from "./firebaseGameUtils";

// const staticPhonicsArray = Object.keys(phonicsData.levelOne);

/**
 * Get a list of phonics from the static array, based on the level.
 * @param {*} level 
 * @returns 
 */
const getStaticPhonicsArray = (level) => {
    let levelData = null;
    switch(level) {
        case 1:
            levelData = "levelOne";
            break;
        case 2:
            levelData = "levelTwo";
            break;
        case 3:
            levelData = "levelThree";
            break;
        case 4:
            levelData = "levelFour";
            break;
        case 5:
            levelData = "levelFive";
            break;
    }
    return Object.keys( phonicsData[levelData] );
}

/**
 * Shuffle that array. Nice and simple algorithm.
 * Code courtesy of Geeks for Geeks.
 * @param {*} array 
 * @returns 
 */
const shuffleArray = (array) => {
    for (var i = array.length - 1; i > 0; i--) {
    
        // Generate random number
        var j = Math.floor(Math.random() * (i + 1));
                    
        var temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
        
    return array;
}
    

/**
 * Get the array of sounds for the swing game.
 * Attempts to return a mix of incorrect, correct and new sounds for the given child, based on their
 * past results. Basic algo at the moment!
 * @param {*} id 
 * @param {*} level 
 * @returns 
 */
export const getArrayForSwing = async (id, level) => {

    // 50 wrong = 8
    // 30 new = 4
    // fill up from rest = 3

    const MAX_PER_GAME = 15;
    const MAX_INCORRECT_PER_GAME = 8;
    const MAX_CORRECT_PER_GAME = 4;

    const game = "swing";

    const staticArray = getStaticPhonicsArray( level );
    console.log(`static array for level ${level}`, staticArray);

    const incorrect = await getArrayOfRounds( id, game, getIncorrectGameRoundsForUser);
    const correct = await getArrayOfRounds( id, game, getCorrectGameRoundsForUser);

    console.log( "correct", correct);
    console.log( "incorrect", incorrect);

    // all results
    const result = new Set();

    // add in at most MAX_INCORRECT_PER_GAME of incorrect answers
    for (let phonic of incorrect) {
        result.add(phonic);
        if( result.length >= MAX_INCORRECT_PER_GAME ) {
            console.log("Break!! Reach max incorrect");
            break;
        }
    }
    console.log("After adding incorrect", result);

    let maxCorrect = result.length + MAX_CORRECT_PER_GAME;
    // add in at most MAX_CORRECT_PER_GAME of correct answers
    for (let phonic of correct) {
        result.add(phonic);
        if( result.length >= maxCorrect ) {
            break;
        }
    }
    console.log("After adding correct", result);

    // fill up remaining
    for (let i = 0; i < staticArray.length; i++) {
        result.add( staticArray[i] );
        if( result.size >= MAX_PER_GAME ) {
            break;
        }
    }

    console.log("At end", result);

    const shuffled = shuffleArray( Array.from(result) );

    return shuffled;
}