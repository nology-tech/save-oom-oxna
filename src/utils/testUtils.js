import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import UserContext from "../contexts/UserContext";

export const customRender = (
  ui,
  useRouting = true,
  contextOptions = { useContext: true, value: undefined }
) => {
  // wrap components in routing and context if requested
  const routingUiResult = useRouting && wrapWithRouting(ui);
  const contextUiResult =
    contextOptions.useContext &&
    wrapWithContext(routingUiResult, contextOptions.value);

  // use RTL's render function to return the test component
  return render(contextUiResult);
};

const wrapWithRouting = (ui) => {
  return <Router>{ui}</Router>;
};

const wrapWithContext = (ui, value) => {
  return <UserContext.Provider value={value}>{ui}</UserContext.Provider>;
};
