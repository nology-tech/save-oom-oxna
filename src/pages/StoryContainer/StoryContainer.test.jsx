import { customRender } from "../../utils/testUtils";
import StoryContainer from "./StoryContainer";

it("Should render the StoryContainer page", () => {
  const { container } = customRender(<StoryContainer />);

  expect(container).toMatchSnapshot();
});
