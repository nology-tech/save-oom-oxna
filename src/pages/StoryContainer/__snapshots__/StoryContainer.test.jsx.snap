// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the StoryContainer page 1`] = `
<div>
  <div
    class="layout"
  >
    <nav
      class="nav-bar"
    >
      <div>
        <a
          href="/dashboard"
        >
          <img
            alt="logo"
            class="logo"
            src="logo.png"
          />
        </a>
        <a
          class="navbar-link"
          href="/swing-game"
        >
          Swing Game
        </a>
        <a
          class="navbar-link"
          href="/dashboard"
        >
          Dashboard
        </a>
        <a
          class="navbar-link"
          href="/stats-page"
        >
          Stats
        </a>
        <h2>
          Barry
        </h2>
      </div>
      <button
        class="button button--secondary"
      >
        Logout
      </button>
    </nav>
    <main
      class="layout__main"
    >
      <div
        class="story-container"
      >
        <p
          class="story-container__text"
        >
          Once upon a time, there was an Alien, whose name was Oom. Oom lived in Space, on planet XXXX.Oom was very excited because when aliens turn 5, they get to fly spaceships for the very first time. Can you see Oom waving? Say Hello to Oom!!
        </p>
        <button
          class="button"
        >
          NEXT
        </button>
      </div>
    </main>
  </div>
</div>
`;
